### DISCUSSION TITLE
Can training be stopped from another process (via callback)?

### DISCUSSION THREAD

**Discussion by haimat** [ultralytics/yolov5#6322] [General]
I would like to control YOLOv5 training from another process (i.e. a custom training GUI), so I am playing with the `utils.callbacks.Callbacks` class utilized in `train.py`. I know that one can get infos on training status and progress via that callback class, however, is there also a way to stop the training process before all epochs are done?

If the `Callbacks` class is not what I am looking for in this context, what other approach should I follow to stop training from another process? I would rather not send a `SIGKILL` signal to the training process ...

**Comment by glenn-jocher** [comment]
@haimat the most frequent callback is probably on_batch_end, so you might put a listener there.

**Reply by haimat** [reply]
Thanks Glenn - but that would mean I need to modify the default `train()` method in `train.py`,, right? It would be great to have that feature without modification, as that would mean I need to modify at every update of YOLOv5. Any ideas?

**Reply by glenn-jocher** [reply]
No, you would add custom code to any of the callbacks like `on_batch_end`.

**Reply by haimat** [reply]
Hmm... I would like to avoid modifying the `train()` function and be able to call it directly from my code. Would such a STOP feature via callbacks (as can be done in Keras callbacks) a suitable feature request?

**Reply by glenn-jocher** [reply]
@haimat yeah of course. I'm not a network engineer so I don't know exactly what the signal handling would look like but I imagine if you just initialize a value and then send a signal to change it with an assert that would be enough, i.e.

```python
# init
kill=False

# update
# update kill value somehow in callback

# check
assert kill=False, 'Run killed'
````

All of the callbacks are here, i.e.:
https://github.com/ultralytics/yolov5/blob/fd55271c04e9be68ea5299f7fe2aafcf4dc1984d/utils/loggers/__init__.py#L83-L97

**Reply by haimat** [reply]
What if we would add a new member `stop_training = False` to the `yolov5.utils.callback.Callbacks` parent class. Then from within the callbacks child class one could simply set `callback.stop_training = True` whenever he wants. And in the `train()` method we could then do this:

```python
callbacks.run('on_train_batch_end', ni, model, imgs, targets, paths, plots, opt.sync_bn)
if callbacks.stop_training:
    return  # or whatever would be best to gracefully stop the training loop
```

Wouldn't that be all we need for this feature?

**Reply by glenn-jocher** [reply]
@haimat yeah that looks like a decent implementation. Can you submit a PR for it?

**Reply by haimat** [reply]
@glenn-jocher Sure, I have created [this PR](https://github.com/ultralytics/yolov5/pull/6365) - is it OK for you?

**Reply by glenn-jocher** [reply]
@haimat great! I will take a look today.

**Reply by haimat** [reply]
@glenn-jocher Don't want to bother you, just asking because I am developing this feature at the moment in our application - can you estimate when you might approve that PR so it will be part of the official master?
